pkg/eventpersister-dwh/persister/evaluation_event.go

```diff
@@ -0,0 +1,230 @@
+		logger: l,
+	}, nil
+}
+
+func (w *evalEvtWriter) Write(
+	ctx context.Context,
+	envEvents environmentEventMap,
+) map[string]bool {
+	evalEvents := []*epproto.EvaluationEvent{}
+	fails := make(map[string]bool, len(envEvents))
+	for environmentNamespace, events := range envEvents {
+		for id, event := range events {
+			switch evt := event.(type) {
+			case *eventproto.EvaluationEvent:
+				e, retriable, err := w.convToEvaluationEvent(ctx, evt, id, environmentNamespace)
+				if err != nil {
+					if err == ErrNoExperiments {
+						w.logger.Warn(
+							"There is no running experiments",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+						continue
+					}
+					if !retriable {
+						w.logger.Error(
+							"Failed to convert to evaluation event",
+							zap.Error(err),
+							zap.String("id", id),
+							zap.String("environmentNamespace", environmentNamespace),
+						)
+					}
+					fails[id] = retriable
+					continue
+				}
+				evalEvents = append(evalEvents, e)
+			default:
+				w.logger.Error(
+					"The event is an unexpected message type",
+					zap.String("id", id),
+					zap.String("environmentNamespace", environmentNamespace),
+				)
+				fails[id] = false
+			}
+		}
+	}
+	if err := w.writer.AppendRows(ctx, evalEvents); err != nil {
+		w.logger.Error(
+			"failed to append rows to evaluation event",
+			zap.Error(err),
+		)
+	}
+	return fails
+}
+
+func (w *evalEvtWriter) convToEvaluationEvent(
+	ctx context.Context,
+	e *eventproto.EvaluationEvent,
+	id, environmentNamespace string,
+) (*epproto.EvaluationEvent, bool, error) {
+	exist, err := w.existExperiment(ctx, e, environmentNamespace)
+	if err != nil {
+		return nil, true, err
+	}
+	if !exist {
+		return nil, false, ErrNoExperiments
+	}
+	if err := w.upsertUserEvaluation(ctx, e, environmentNamespace); err != nil {
+		return nil, true, err
+	}
+	var ud []byte
+	if e.User != nil {
+		var err error
+		ud, err = json.Marshal(e.User.Data)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+	tag := e.Tag
+	if tag == "" {
+		// For requests with no tag, it will insert "none" instead, until all old SDK clients are updated
+		tag = "none"
+	}
+	return &epproto.EvaluationEvent{
+		Id:                   id,
+		FeatureId:            e.FeatureId,
+		FeatureVersion:       e.FeatureVersion,
+		UserData:             string(ud),
+		UserId:               e.UserId,
+		VariationId:          e.VariationId,
+		Reason:               e.Reason.Type.String(),
+		Tag:                  tag,
+		SourceId:             e.SourceId.String(),
+		EnvironmentNamespace: environmentNamespace,
+		Timestamp:            e.Timestamp,
+	}, false, nil
+}
+
+func (w *evalEvtWriter) existExperiment(
+	ctx context.Context,
+	event *eventproto.EvaluationEvent,
+	environmentNamespace string,
+) (bool, error) {
+	resp, err := w.experimentClient.ListExperiments(ctx, &exproto.ListExperimentsRequest{
```

Comment:
I know I agreed to check for running experiments using the `FeatureId` and `FeatureVersion`, but giving second thought, it will be better not to use them.

Because we usually receive more EvaluationEvent than GoalEvent, it will make a lot of requests in the Experiment service. For that reason, it is better if we use singleflight and list all the running experiments, so the same requests can be shared between the batches as we do in the [GoalEvent](https://github.com/bucketeer-io/bucketeer/pull/150/files#diff-6fe94d1ec57d527e5c43382abd5801cd6f0f5ec7ae992ec5cce5d3419a2f65a1R232-R266).